### PR TITLE
Add decorator to route to provide additional information, including t…

### PR DESCRIPTION
**Pull request by objectiser**
…he operation for use with the tracing span.

Related to new Decorator message in https://github.com/lyft/envoy-api/pull/171

### CODE REVIEW

File: test/common/tracing/http_tracer_impl_test.cc [hunk 1]
```diff
@@ -442,24 +442,20 @@ class HttpTracerImplTest : public Test {
 };
 
 TEST_F(HttpTracerImplTest, BasicFunctionalityNullSpan) {
-  EXPECT_CALL(config_, operationName()).Times(2);
+  EXPECT_CALL(config_, operationName()).Times(1);
```
Comment (htuch) [inline test/common/tracing/http_tracer_impl_test.cc]: Curious why this changes..

File: examples/jaeger-tracing/front-envoy-jaeger.json [hunk 2]
```diff
@@ -20,7 +20,10 @@
                   "routes": [
                     {
                       "prefix": "/",
-                      "cluster": "service1"
+                      "cluster": "service1",
+                      "decorator": {
```
Comment (RomanDzhabarov) [inline examples/jaeger-tracing/front-envoy-jaeger.json]: `decorator` seems to be pretty generic name, should it be tracing_decorator. And also we need doc update on this.

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -76,6 +76,15 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers)
   return hash;
 }
 
+DecoratorImpl::DecoratorImpl(const envoy::api::v2::Decorator& decorator)
+    : operation_(decorator.operation()) {}
+
+void DecoratorImpl::apply(Tracing::Span& span) const {
+  if (operation_.size() > 0) {
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: isEmpty

File: include/envoy/router/router.h [hunk 4]
```diff
@@ -11,6 +11,7 @@
 #include "envoy/common/optional.h"
 #include "envoy/http/codec.h"
 #include "envoy/http/header_map.h"
+#include "envoy/tracing/http_tracer.h"
```
Comment (RomanDzhabarov) [inline include/envoy/router/router.h]: why is this included?

File: source/common/tracing/http_tracer_impl.cc [hunk 5]
```diff
@@ -168,11 +168,10 @@ HttpTracerImpl::HttpTracerImpl(DriverPtr&& driver, const LocalInfo::LocalInfo& l
 
 SpanPtr HttpTracerImpl::startSpan(const Config& config, Http::HeaderMap& request_headers,
                                   const Http::AccessLog::RequestInfo& request_info) {
-  std::string span_name = HttpTracerUtility::toString(config.operationName());
+  std::string span_name;
```
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: what is the span name we are going to report for the ingress?

File: test/common/tracing/http_tracer_impl_test.cc (hunk 1 above)
Comment (objectiser) [reply test/common/tracing/http_tracer_impl_test.cc]: In `HttpTracerImpl::startSpan` I removed the initialisation of the span name based on the `operationName()`, as it was not being used in the ZipkinTracer (it was overriding the span name to use the host). I need to check with someone from LightStep to see if it has an impact on their tracer - from initial inspection it seems like the span name is just passed through to the tracer, so I'm guessing Igress/Egress would just appear as the span names in their UI. However this is not that meaningful, as the span.kind can be used to indicate direction.

However this is something that needs to be checked before this PR is merged.

File: source/common/tracing/http_tracer_impl.cc (hunk 5 above)
Comment (objectiser) [reply source/common/tracing/http_tracer_impl.cc]: For zipkin tracer, this was not set - so only change is that the Ingress value would not be provided to LightStep's tracer - so before this PR is merged this needs to be checked with them.

File: examples/jaeger-tracing/front-envoy-jaeger.json (hunk 2 above)
Comment (objectiser) [reply examples/jaeger-tracing/front-envoy-jaeger.json]: Using the more general `decorator` term was @mattklein123 's idea: https://github.com/lyft/envoy-api/pull/171#discussion_r138397108

Will sort out docs soon.

File: include/envoy/router/router.h (hunk 4 above)
Comment (objectiser) [reply include/envoy/router/router.h]: Its required for the `apply(Tracing::Span&)` on `Decorator`.

File: source/common/tracing/http_tracer_impl.cc (hunk 5 above)
Comment (objectiser) [reply source/common/tracing/http_tracer_impl.cc]: @bhs Could you provide some guidance here?

Prior to this PR, this code was building a `span_name` based on the 'OperationName` (misleading) which essentially just represents the direction of the request (Ingress or Egress), and if Egress, then it appends the host. However the Zipkin tracer code was ignoring this value, and instead using the host for Egress requests and empty span name for Ingress.

It appears that the `span_name` is just being passed through to the LightStep tracer and not doing any kind of interpretation of the value to determine span kind. So assume this value is just being included in the LightStep spans and displayed via your UI?

If so, do you have any objections to the changed values (i.e. blank on Ingress and host for Egress), or would you prefer different default values? With this PR, if the configuration includes a specified `operation` for the route, then this would be used instread.

**Comment by htuch** [comment]
@objectiser I'm going to fix it so we consume your envoy-api change now to unblock this PR (and help with one I'm putting together), so should be done Real Soon.

File: source/common/tracing/http_tracer_impl.cc (hunk 5 above)
Comment (RomanDzhabarov) [reply source/common/tracing/http_tracer_impl.cc]: LS does not care about `operation name`, but if affects `Saved Searches` if there are any. Do we really need to change this behavior?
It seems like if there is a decorator on a route, span name will be overwritten and if not, that this default one is going to be used.

Comment (objectiser) [reply source/common/tracing/http_tracer_impl.cc]: That is the problem - the default behaviour is different between the two tracers. LS uses the default value passed in, but the Zipkin tracer overrides the supplied value and just uses the Host for Egress routes.

So not sure how to allow both tracers to maintain their current default behaviour, while at the same time enabling the Decorator to override the span name in both.

**Comment by objectiser** [comment]
@htuch @mattklein123 rebased off envoy master to make use of the latest data-plane-api.

File: test/common/tracing/http_tracer_impl_test.cc (hunk 1 above)
Comment (htuch) [reply test/common/tracing/http_tracer_impl_test.cc]: Any update on this? PR looks LGTM otherwise.

Comment (objectiser) [reply test/common/tracing/http_tracer_impl_test.cc]: I've been in contact with a couple of people at LightStep who said they would try to look at it, but not sure when.

Comment (objectiser) [reply test/common/tracing/http_tracer_impl_test.cc]: cc @bhs @tedsuo @rnburn Any thoughts appreciated - would ideally like to get this PR merged, so just need to understand if a change in the operation name provided to the LightStep tracer would have any impact on your users?

File: source/common/tracing/http_tracer_impl.cc (hunk 5 above)
Comment (RomanDzhabarov) [reply source/common/tracing/http_tracer_impl.cc]: Got it, if there are no common names for spans for LS and Zipkin spans, I'd push the logic of assigning specific span names to the specific tracer (ls and zipkin tracers).

File: include/envoy/router/router.h [hunk 6]
```diff
@@ -261,6 +262,26 @@ class RouteEntry {
...
+   * @param Tracing::Span& the span.
+   */
+  virtual void apply(Tracing::Span& span) const PURE;
+};
+
+/**
+ * These Decorator shared pointers are created on the main thread at config ingest time
+ * and then shared with worker threads as spans are processed.
+ */
+typedef std::shared_ptr<const Decorator> DecoratorConstSharedPtr;
```
Comment (mattklein123) [inline include/envoy/router/router.h]: You should be able to store the decorator in a unique_ptr.

File: source/common/tracing/http_tracer_impl.cc (hunk 5 above)
Comment (mattklein123) [reply source/common/tracing/http_tracer_impl.cc]: IMO the LS current behavior makes more sense. Meaning, if no operation is set, base operation name off of egress/ingress + host, etc. If the tracer wants to override (as you do inside Zipkin) that seems fine? It doesn't make much sense to me to push this logic entirely into each tracer, but I suppose we could do that. Maybe I'm not fully following the issue here.

File: test/common/tracing/http_tracer_impl_test.cc (hunk 1 above)
Comment (mattklein123) [reply test/common/tracing/http_tracer_impl_test.cc]: Per other comment and per @RomanDzhabarov yes it effects the UI. I'm not sure why we need to change the default behavior. If no operation name is specified, do what we do today. If operation name is specified, override it. If tracer wants to override that, no problem.

Also nit: Times(1) is redundant.

File: source/common/tracing/http_tracer_impl.cc (hunk 5 above)
Comment (objectiser) [reply source/common/tracing/http_tracer_impl.cc]: @mattklein123 @RomanDzhabarov You are correct - I was still thinking about the previous way I had implemented it, where the tracing config provided operation was passed in when the span was started - whereas now it is set after the span has been created within the Decorator - so all good, I can return the default behaviour to previous versions.

File: docs/configuration/http_conn_man/route_config/route.rst [hunk 7]
```diff
@@ -406,6 +411,24 @@ header_name
...
+Specifies the route's decorator.
+
+.. code-block:: json
+
+   {
+     "operation": "..."
+   }
+
+operation
+  *(required, string)* The operation name associated with the request matched to this route. If tracing is
```
Comment (RomanDzhabarov) [inline docs/configuration/http_conn_man/route_config/route.rst]: nit: will be used as a span name?

File: examples/jaeger-tracing/front-envoy-jaeger.json (hunk 2 above)
Comment (RomanDzhabarov) [reply examples/jaeger-tracing/front-envoy-jaeger.json]: fair enough

**Comment by objectiser** [comment]
@htuch @mattklein123 As the CLA approach has been changed to use the git commit sign off approach, should I squash the commits and just sign the single commit?

**Comment by mattklein123** [comment]
@objectiser yeah sorry for trouble just squash/rebase/push with DCO.

**Comment by objectiser** [comment]
@mattklein123 No problem, we doing the same change in Jaeger.

File: source/common/router/config_impl.cc [hunk 8]
```diff
@@ -255,6 +264,14 @@ RouteEntryImplBase::parseOpaqueConfig(const envoy::api::v2::Route& route) {
   return ret;
 }
 
+DecoratorConstPtr RouteEntryImplBase::parseDecorator(const envoy::api::v2::Route& route) {
+  DecoratorConstPtr ret;
+  if (route.has_decorator()) {
+    ret = std::unique_ptr<Decorator>(new DecoratorImpl(route.decorator()));
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: nit: `DecoratorConstPtr(...`

File: source/common/http/conn_manager_impl.cc [hunk 9]
```diff
@@ -543,6 +543,9 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
 
     if (tracing_decision.is_tracing) {
       active_span_ = connection_manager_.tracer_.startSpan(*this, *request_headers_, request_info_);
+      if (cached_route_.value() && cached_route_.value()->decorator()) {
+        cached_route_.value()->decorator()->apply(*active_span_);
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc]: I think you will need a conn_manager_impl_test to get test coverage for this line.

File: source/common/tracing/lightstep_tracer_impl.cc [hunk 10]
```diff
@@ -23,6 +23,10 @@ void LightStepSpan::finishSpan(SpanFinalizer& finalizer) {
   span_.Finish();
 }
 
+void LightStepSpan::setOperation(const std::string& operation) {
+  span_.SetOperationName(operation);
```
Comment (mattklein123) [inline source/common/tracing/lightstep_tracer_impl.cc]: Does this need coverage in lightstep tracer?

Comment (objectiser) [reply source/common/tracing/lightstep_tracer_impl.cc]: Not sure best way to test this - the current lightstep tests don't check the operation names passed in via the `startSpan` method, and there are no accessors for the operation name. Any suggestions welcome.

Comment (mattklein123) [reply source/common/tracing/lightstep_tracer_impl.cc]: If it's not easy to verify it's fine just skip it. I would at least make sure it's called just to have basic crash avoidance testing. (It might already be called not sure).

**Comment by mattklein123** [comment]
@RomanDzhabarov any final comments?
